src/embedded-checkout/embedded-checkout-listener.ts

```diff
@@ -0,0 +1,75 @@
+import { bindDecorator as bind } from '../common/utility';
+
+import { EmbeddedCheckoutEvent, EmbeddedCheckoutEventMap } from './embedded-checkout-events';
+import isEmbeddedCheckoutEvent from './is-embedded-checkout-event';
+
+export default class EmbeddedCheckoutListener {
+    private _isListening: boolean;
+    private _listeners: {
+        [key: string]: Array<(event: EmbeddedCheckoutEventMap[keyof EmbeddedCheckoutEventMap]) => void>;
+    };
+
+    constructor(
+        private _origin: string
+    ) {
+        this._isListening = false;
+        this._listeners = {};
+    }
+
+    listen(): void {
+        if (this._isListening) {
+            return;
+        }
+
+        this._isListening = true;
+
+        window.addEventListener('message', this._handleMessage);
+    }
+
+    stopListen(): void {
+        if (!this._isListening) {
+            return;
+        }
+
+        this._isListening = false;
+
+        window.removeEventListener('message', this._handleMessage);
+    }
+
+    addListener<TType extends keyof EmbeddedCheckoutEventMap>(type: TType, listener: (event: EmbeddedCheckoutEventMap[TType]) => void): void {
+        if (!this._listeners[type]) {
+            this._listeners[type] = [];
+        }
+
+        this._listeners[type].push(listener);
+    }
+
+    removeListener<TType extends keyof EmbeddedCheckoutEventMap>(type: TType, listener: (event: EmbeddedCheckoutEventMap[TType]) => void): void {
+        const index = this._listeners[type] && this._listeners[type].indexOf(listener);
```

Comment:
if `this._listeners[type]` is undefined, would this still try to splice? Maybe just add a test that tries to remove a listener from a type that doesnt exist?